Prepare for dart 3.0 alpha changes: generate assets

annagrin opened the pull request:
- Prepare or Dart 3 alpha breaking changes:
  - Generate missing SDK assets for tests.
  - Enable null safe tests that use the generated assets.

Towards: https://github.com/dart-lang/webdev/issues/1878
Helps: https://github.com/dart-lang/webdev/issues/1818
Closes: https://github.com/dart-lang/webdev/issues/1591

nshahan commented on dwds/lib/src/utilities/sdk_asset_generator.dart:
```diff
@@ -0,0 +1,221 @@
+      // we only create one for weak mode.
+      // This creates full dill file in '$fullDillPath'.
+      if (!hasFullDillAsset) '--summarize',
+      if (soundNullSafety) '--sound-null-safety',
+      if (!soundNullSafety) '--no-sound-null-safety',
```
I'm just curious about using two if-elements vs using one with an else clause. Is this on purpose?

annagrin replied:
I am removing the conditional summarizing in following updates.

annagrin replied:
ah i din't know you could have an `else` here:) will change!

annagrin replied:
done!

nshahan commented on dwds/lib/src/utilities/sdk_asset_generator.dart:
```diff
@@ -0,0 +1,221 @@
+      '--source',
+      'dart:core',
+      '--summary-only',
+      if (soundNullSafety) '--sound-null-safety',
+      if (!soundNullSafety) '--no-sound-null-safety',
```
ditto

annagrin replied:
ack

nshahan commented on dwds/lib/src/utilities/sdk_configuration.dart:
```diff
@@ -85,14 +89,23 @@ class SdkConfiguration {
   }
 
   void validateSummaries({FileSystem fileSystem = const LocalFileSystem()}) {
+    validateWeakSummaries(fileSystem: fileSystem);
+    validateWeakSummaries(fileSystem: fileSystem);
```
Same line twice, copy-paste typo?

annagrin replied:
oops:) thanks!

nshahan commented on dwds/lib/src/utilities/sdk_configuration.dart:
```diff
@@ -116,27 +129,64 @@ class SdkConfiguration {
+/// Implementation for SDK configuration for tests that can generate missing assets.
+/// Frontend server:
+///  - need to generate SDK js (normally included in flutter SDK).
+///  - need to generate SDK summary for weak null safety mode as it is not
+///    provided by the SDK installation.
```
Is this actually a summary or the full .dill?

annagrin replied:
In weak mode we need to generate everything, will update.

annagrin replied:
updated.

nshahan commented on dwds/lib/src/utilities/sdk_asset_generator.dart:
```diff
@@ -0,0 +1,221 @@
+  final String weakAssetDirectory;
+  final String dartSdkJsDirectory;
+
+  String get soundSummaryPath =>
+      p.join(soundAssetDirectory, 'ddc_outline_sound.dill');
+  String get soundFullDillPath =>
+      p.join(soundAssetDirectory, 'ddc_platform_sound.dill');
+  String get soundJsPath => p.join(dartSdkJsDirectory, _sdkJsSoundFileName);
+  String get soundJsMapPath =>
+      p.join(dartSdkJsDirectory, _sdkJsMapSoundFileName);
+
+  String get weakSummaryPath => p.join(weakAssetDirectory, 'ddc_sdk.dill');
+  String get weakFullDillPath =>
+      p.join(weakAssetDirectory, 'ddc_platform.dill');
```
Same here, if these names are completely under your control I would use the following for consistency and ease of deletion in the future.

ddc_outline.dill (sound outline dill)
ddc_platform.dill (sound full dill)
ddc_outline_unsound.dill (weak/unsound outline dill)
ddc_platform_unsound.dill (weak/unsound full dill)

annagrin replied:
Those are the current names in the sdk - will change them to match the new namew in the next PR.

nshahan commented on webdev/test/test_utils.dart:
```diff
@@ -45,3 +49,43 @@ Map<String, String> getPubEnvironment() {
+  /// Configuration matching the default SDK layout.
+  static final defaultSdkConfiguration = SdkConfiguration(
+    sdkDirectory: sdkDir,
+    unsoundSdkSummaryPath: p.join(sdkDir, 'lib', '_internal', 'ddc_sdk.dill'),
+    soundSdkSummaryPath:
+        p.join(sdkDir, 'lib', '_internal', 'ddc_outline_sound.dill'),
```
Are these paths reading from the SDK or are you writing those .dill files? If coming from the SDK, we will need to update these very soon when I land my changes.

annagrin commented on dwds/lib/src/utilities/sdk_asset_generator.dart:
```diff
@@ -0,0 +1,221 @@
+  static bool _sdkAssetsGenerated = false;
+  final _logger = Logger('SdkAssetGenerator');
+
+  final _sdkJsWeakFileName = 'dart_sdk.js';
+  final _sdkJsMapWeakFileName = 'dart_sdk.js.map';
+  final _sdkJsSoundFileName = 'dart_sdk_sound.js';
+  final _sdkJsMapSoundFileName = 'dart_sdk_sound.js.map';
```
Those are the current names in the sdk - will change them to match the new namew in the next PR.

annagrin commented on dwds/lib/src/utilities/sdk_asset_generator.dart:
```diff
@@ -0,0 +1,221 @@
+import 'package:logging/logging.dart';
+import 'package:path/path.dart' as p;
+
+/// Generates sdk.js, sdk.map, and sdk summary files.
```
Updated the comments, thanks!

annagrin commented on dwds/lib/src/utilities/sdk_asset_generator.dart:
```diff
@@ -0,0 +1,221 @@
+/// Missing assets:
+/// mode\env | test (FE server) | test (build daemon) | test (webdev)
+/// -----------------------------------------------------------------
+/// strong   | js               |                     |
```
done!

elliette commented on dwds/lib/src/utilities/sdk_asset_generator.dart:
```diff
@@ -0,0 +1,244 @@
+import 'dart:convert';
```
Is this only used for tests? (Should it be in the test directory)?

annagrin replied:
webdev has to use it in its tests as well so I have to export it unfortunately....

jakemac53 replied:
You could also create a separate package, with testing utils, that isn't published on pub.

annagrin replied:
will do!

annagrin replied:
Actually, I would like to do this in a separate PR as it will require some other files to move to the _test_common and a lot of changes in tests. I removed exporting the generator already but kept it in the lib/src  for this PR.

elliette replied:
Could we move this to the `dwds/test/fixtures` directory for now and remove the webdev test changes from this PR? In the follow-up PR creating `_test_common` we can add back in those webdev test changes.

annagrin replied:
sure

annagrin replied:
done!

elliette commented on dwds/lib/dwds.dart:
```diff
@@ -29,4 +29,5 @@ export 'src/services/expression_compiler.dart'
 export 'src/utilities/sdk_configuration.dart'
-    show SdkConfiguration, SdkConfigurationProvider;
+    show SdkLayout, SdkConfiguration, SdkConfigurationProvider;
+export 'src/utilities/sdk_asset_generator.dart' show SdkAssetGenerator;
```
Why is this being exported?

annagrin replied:
webdev has to use it in its tests as well so I have to export it unfortunately....

annagrin replied:
Alternatively, I can have a copy of it in webdev - maybe it is better than exporting?

jakemac53 replied:
You can also import from `src` inside webdev. We control both packages so you don't have to worry much about breakage, especially if its only inside tests.

annagrin replied:
> You can also import from src inside webdev. 
cool, will try that, thanks!

annagrin replied:
That worked! Updated.

nshahan reviewed (approved):
LGTM pending the comments from @elliette.

annagrin commented on dwds/lib/src/utilities/sdk_asset_generator.dart:
```diff
@@ -0,0 +1,244 @@
+
+/// Generates sdk.js, sdk.map, sdk full dill, and sdk summary files.
+///
+/// Missing assets:
```
will do!

annagrin replied:
Done, simplified it as well.

elliette commented on dwds/CHANGELOG.md:
```diff
@@ -17,13 +17,18 @@
 - Include an optional param to `Dwds.start` to indicate whether it a Flutter app
   or not.
 - Remove deprecated `ChromeProxyService.setExceptionPauseMode()`.
+- Replace `SdkConfiguration.unsoundSdkSummaryPath` by 
```
Nit: by -> with

annagrin replied:
done!

elliette commented:
I don't think the `mono_pkg.yaml` / `mono repo` changes should be part of this PR. It looks like we are missing `mono_pkg.yaml` files in some of our test fixtures, I think we should add those separately.

elliette reviewed (approved):
LGTM with a few remaining comments. Thanks!

elliette commented:
> I don't think the `mono_pkg.yaml` / `mono repo` changes should be part of this PR. It looks like we are missing `mono_pkg.yaml` files in some of our test fixtures, I think we should add those separately.

Ah, remembered those changes are being tracked here: https://github.com/dart-lang/webdev/issues/1845

annagrin commented on dwds/CHANGELOG.md:
```diff
@@ -17,13 +17,18 @@
 - Add back `ChromeProxyService.setExceptionPauseMode()` without override.
 - Make hot restart atomic to prevent races on simultaneous execution.
 - Return error on expression evaluation if expression evaluator stopped.
+- Prepare or Dart 3 alpha breaking changes:
```
done!

annagrin commented:
> > I don't think the `mono_pkg.yaml` / `mono repo` changes should be part of this PR. It looks like we are missing `mono_pkg.yaml` files in some of our test fixtures, I think we should add those separately.
> 
> Ah, remembered those changes are being tracked here: #1845

Since I had to move webdev tests to sound nullsafety, it does not seem reasonable not not run the analyzer on the same repo. So I updated it as well (is also fixed an analyzer test breaking)
